In [2]:
acc1 <- read.csv('acc1_final.csv')

In [3]:
head(acc1,10)

X gid        accident_cnt 어린이집_count 유치원_count 초등학교_count
1  0 다사561083 1            0              0            0             
2  1 다사564096 1            0              0            0             
3  2 다사566102 1            0              0            0             
4  3 다사583093 1            0              0            0             
5  4 다사584072 1            0              0            0             
6  5 다사585098 1            2              1            0             
7  6 다사586106 1            8              0            0             
8  7 다사587044 1            0              0            0             
9  8 다사587072 1            1              0            0             
10 9 다사587098 1            0              1            0             
   방지턱_count 불법주정차_cam_count 정류장_count_100 정류장_count_300 ⋯
1  0            0                    2                 2               ⋯
2  0            0                    0                 2               ⋯
3  0            0                    0                 2               ⋯
4  0            0                    0                 0               ⋯
5  0            0                    1                 6               ⋯
6  0            0                    0                 0               ⋯
7  0            0                    0                10               ⋯
8  0            0                    0                 3               ⋯
9  0            0                    0                 6               ⋯
10 0            0                    0                 3               ⋯
   어린이집_count_100 유치원_count_100 초등학교_count_100 방지턱_count_100
1  0                  0                0                  0               
2  0                  0                0                  0               
3  0                  0                0                  0               
4  0                  0                0                  0               
5  0                  0                0                  0               
6  0                  0                0                  0               
7  0                  0                0                  0               
8  0                  0                0                  0               
9  0                  0                0                  0               
10 0                  0                0                  0               
   grid_유소년인구_100 grid_학원수_100 어린이집_count_200 유치원_count_200
1   0                  0               0                  0               
2   0                  0               0                  0               
3   0                  0               0                  0               
4   0                  0               0                  0               
5   0                  0               0                  0               
6  30                  0               0                  1               
7  81                  0               4                  0               
8   0                  0               0                  0               
9   0                  0               1                  0               
10  0                  0               0                  1               
   초등학교_count_200 방지턱_count_200
1  0                  0               
2  0                  0               
3  0                  0               
4  0                  0               
5  0                  0               
6  0                  0               
7  0                  0               
8  0                  0               
9  0                  0               
10 0                  0

In [4]:
str(acc1)

'data.frame':	345 obs. of  24 variables:
 $ X                   : int  0 1 2 3 4 5 6 7 8 9 ...
 $ gid                 : Factor w/ 345 levels "다사561083","다사564096",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ accident_cnt        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ 어린이집_count      : int  0 0 0 0 0 2 8 0 1 0 ...
 $ 유치원_count        : int  0 0 0 0 0 1 0 0 0 1 ...
 $ 초등학교_count      : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 방지턱_count        : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 불법주정차_cam_count: int  0 0 0 0 0 0 0 0 0 0 ...
 $ 정류장_count_100    : int  2 0 0 0 1 0 0 0 0 0 ...
 $ 정류장_count_300    : int  2 2 2 0 6 0 10 3 6 3 ...
 $ grid_유동인구       : num  43.436 9.633 0.215 0.662 8.104 ...
 $ grid_유소년인구     : num  0 0 0 0 0 30 411 6 0 11 ...
 $ grid_학원수         : int  0 0 0 0 0 0 8 0 0 0 ...
 $ geometry            : Factor w/ 345 levels "MULTIPOLYGON (((127.0054916936862 37.17242713739633, 127.005485819363 37.17332852036621, 127.0066122372272 37.1"| __truncated__,..: 1 2 3 4 5 6 7 11 10 9 ...
 $ 어린이집_count_100  : int  0 0 

In [5]:
# 필요없는 열 제거 > X, gid, geometry
acc1 <- subset(acc1, select = -c(X,gid,geometry))

In [6]:
str(acc1)

'data.frame':	345 obs. of  21 variables:
 $ accident_cnt        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ 어린이집_count      : int  0 0 0 0 0 2 8 0 1 0 ...
 $ 유치원_count        : int  0 0 0 0 0 1 0 0 0 1 ...
 $ 초등학교_count      : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 방지턱_count        : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 불법주정차_cam_count: int  0 0 0 0 0 0 0 0 0 0 ...
 $ 정류장_count_100    : int  2 0 0 0 1 0 0 0 0 0 ...
 $ 정류장_count_300    : int  2 2 2 0 6 0 10 3 6 3 ...
 $ grid_유동인구       : num  43.436 9.633 0.215 0.662 8.104 ...
 $ grid_유소년인구     : num  0 0 0 0 0 30 411 6 0 11 ...
 $ grid_학원수         : int  0 0 0 0 0 0 8 0 0 0 ...
 $ 어린이집_count_100  : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 유치원_count_100    : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 초등학교_count_100  : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 방지턱_count_100    : int  0 0 0 0 0 0 0 0 0 0 ...
 $ grid_유소년인구_100 : num  0 0 0 0 0 30 81 0 0 0 ...
 $ grid_학원수_100     : int  0 0 0 0 0 0 0 0 0 0 ...
 $ 어린이집_count_200  : int  0 0 0 0 0 0 4 0 1 0 ...
 $ 유치원_count_200    : int  0 0 0

In [7]:
summary(acc1)

  accident_cnt    어린이집_count    유치원_count   초등학교_count  
 Min.   : 1.000   Min.   : 0.000   Min.   :0.000   Min.   :0.0000  
 1st Qu.: 1.000   1st Qu.: 1.000   1st Qu.:0.000   1st Qu.:0.0000  
 Median : 1.000   Median : 4.000   Median :1.000   Median :1.0000  
 Mean   : 1.557   Mean   : 5.429   Mean   :1.026   Mean   :0.6203  
 3rd Qu.: 2.000   3rd Qu.: 8.000   3rd Qu.:2.000   3rd Qu.:1.0000  
 Max.   :10.000   Max.   :23.000   Max.   :4.000   Max.   :2.0000  
  방지턱_count    불법주정차_cam_count 정류장_count_100 정류장_count_300
 Min.   :0.0000   Min.   : 0.000       Min.   :0.000    Min.   : 0.000  
 1st Qu.:0.0000   1st Qu.: 0.000       1st Qu.:0.000    1st Qu.: 4.000  
 Median :0.0000   Median : 2.000       Median :1.000    Median : 7.000  
 Mean   :0.7507   Mean   : 2.612       Mean   :1.003    Mean   : 6.823  
 3rd Qu.:1.0000   3rd Qu.: 4.000       3rd Qu.:2.000    3rd Qu.: 9.000  
 Max.   :7.0000   Max.   :11.000       Max.   :5.000    Max.   :19.000  
 grid_유동인구    grid_유소년인구   grid_학원수   

In [10]:
library(car)
vif(lm_fit)

어린이집_count         유치원_count       초등학교_count 
            6.081013             3.777740             3.473446 
        방지턱_count 불법주정차_cam_count     정류장_count_100 
            2.609875             1.956711             1.281043 
    정류장_count_300        grid_유동인구      grid_유소년인구 
            1.405075             1.458284             5.923739 
         grid_학원수   어린이집_count_100     유치원_count_100 
            2.417858             2.093999             2.456654 
  초등학교_count_100     방지턱_count_100  grid_유소년인구_100 
            2.355160             1.524568             3.466117 
     grid_학원수_100   어린이집_count_200     유치원_count_200 
            1.445737             6.753024             5.202518 
  초등학교_count_200     방지턱_count_200 
            4.537598             3.016714

- 다중공선성 전부 10 미만
- 변수가 선형종속 약함

### Fitting Linear Model

In [10]:
lm_fit = lm(accident_cnt~.,acc1)
summary(lm_fit)


Call:
lm(formula = accident_cnt ~ ., data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.3998 -0.5608 -0.2970  0.2472  7.8299 

Coefficients:
                       Estimate Std. Error t value Pr(>|t|)    
(Intercept)           1.1843638  0.1465220   8.083 1.27e-14 ***
어린이집_count        0.0452167  0.0278152   1.626   0.1050    
유치원_count          0.1163800  0.1183207   0.984   0.3260    
초등학교_count       -0.0081233  0.1600085  -0.051   0.9595    
방지턱_count          0.0192053  0.0798451   0.241   0.8101    
불법주정차_cam_count -0.0034368  0.0332635  -0.103   0.9178    
정류장_count_100      0.1503724  0.0586619   2.563   0.0108 *  
정류장_count_300     -0.0044460  0.0194273  -0.229   0.8191    
grid_유동인구         0.0017955  0.0018279   0.982   0.3267    
grid_유소년인구      -0.0001490  0.0006407  -0.233   0.8162    
grid_학원수           0.0005944  0.0026808   0.222   0.8247    
어린이집_count_100   -0.0812489  0.0806626  -1.007   0.3146    
유치원_count_100     -0.0139077  0.3028168  -0.046  

In [11]:
step(lm_fit,direction = 'both')

Start:  AIC=75.64
accident_cnt ~ 어린이집_count + 유치원_count + 초등학교_count + 
    방지턱_count + 불법주정차_cam_count + 정류장_count_100 + 
    정류장_count_300 + grid_유동인구 + grid_유소년인구 + 
    grid_학원수 + 어린이집_count_100 + 유치원_count_100 + 
    초등학교_count_100 + 방지턱_count_100 + grid_유소년인구_100 + 
    grid_학원수_100 + 어린이집_count_200 + 유치원_count_200 + 
    초등학교_count_200 + 방지턱_count_200

                       Df Sum of Sq    RSS    AIC
- 유치원_count_100      1    0.0025 380.34 73.647
- 초등학교_count        1    0.0030 380.34 73.647
- 방지턱_count_100      1    0.0058 380.35 73.650
- 불법주정차_cam_count  1    0.0125 380.35 73.656
- 방지턱_count_200      1    0.0202 380.36 73.663
- 어린이집_count_200    1    0.0248 380.36 73.667
- grid_학원수           1    0.0577 380.40 73.697
- 정류장_count_300      1    0.0615 380.40 73.700
- grid_유소년인구       1    0.0635 380.40 73.702
- 방지턱_count          1    0.0679 380.41 73.706
- 초등학교_count_100    1    0.3519 380.69 73.964
- grid_유소년인구_100   1    0.7784 381.12 74.350
- grid_유동인구         1    1.1326 3


Call:
lm(formula = accident_cnt ~ 어린이집_count + 정류장_count_100 + 
    어린이집_count_100 + grid_유소년인구_100 + grid_학원수_100, 
    data = acc1)

Coefficients:
        (Intercept)       어린이집_count     정류장_count_100  
           1.224843             0.049207             0.155418  
 어린이집_count_100  grid_유소년인구_100      grid_학원수_100  
          -0.109778            -0.001495             0.010592  


In [16]:
lm_fit2 <- lm(accident_cnt ~ 어린이집_count + 정류장_count_100  + 어린이집_count_100 + grid_유소년인구_100 + grid_학원수_100,acc1)
summary(lm_fit2)


Call:
lm(formula = accident_cnt ~ 어린이집_count + 정류장_count_100 + 
    어린이집_count_100 + grid_유소년인구_100 + grid_학원수_100, 
    data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.2708 -0.5601 -0.2793  0.2844  7.9499 

Coefficients:
                      Estimate Std. Error t value Pr(>|t|)    
(Intercept)          1.2248428  0.0979307  12.507  < 2e-16 ***
어린이집_count       0.0492068  0.0148481   3.314  0.00102 ** 
정류장_count_100     0.1554181  0.0518186   2.999  0.00291 ** 
어린이집_count_100  -0.1097782  0.0689020  -1.593  0.11204    
grid_유소년인구_100 -0.0014953  0.0008237  -1.815  0.07034 .  
grid_학원수_100      0.0105922  0.0049851   2.125  0.03433 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.07 on 339 degrees of freedom
Multiple R-squared:  0.08306,	Adjusted R-squared:  0.06954 
F-statistic: 6.142 on 5 and 339 DF,  p-value: 1.829e-05


- 어린이집_count_100 유의성x,  grid_유소년인구_100은 조금 벗어났지만 제거하지 않기로

In [19]:
# 어린이집_count_100 제거
lm.final <- lm(accident_cnt ~ 어린이집_count + 정류장_count_100  + grid_유소년인구_100 + grid_학원수_100,acc1)
summary(lm.final)


Call:
lm(formula = accident_cnt ~ 어린이집_count + 정류장_count_100 + 
    grid_유소년인구_100 + grid_학원수_100, data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.3727 -0.5669 -0.3100  0.2725  7.8574 

Coefficients:
                     Estimate Std. Error t value Pr(>|t|)    
(Intercept)          1.234432   0.097966  12.601  < 2e-16 ***
어린이집_count       0.038373   0.013230   2.901  0.00397 ** 
정류장_count_100     0.160065   0.051853   3.087  0.00219 ** 
grid_유소년인구_100 -0.001757   0.000809  -2.172  0.03057 *  
grid_학원수_100      0.010582   0.004996   2.118  0.03490 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.072 on 340 degrees of freedom
Multiple R-squared:  0.07619,	Adjusted R-squared:  0.06532 
F-statistic: 7.011 on 4 and 340 DF,  p-value: 1.965e-05


- 결정계수 0.07 / 다른 계수들은 유의함
- 종속변수가 곡선이기에, 비선형 regression model을 fitting

In [21]:
# (1) 종속변수 logarithm applying / 설명변수 2차항까지 fitting
lm.final2 <- lm(log(accident_cnt) ~ poly(어린이집_count,2) + poly(정류장_count_100,2)  + poly(grid_유소년인구_100,2) + poly(grid_학원수_100,2),acc1)
summary(lm.final2)


Call:
lm(formula = log(accident_cnt) ~ poly(어린이집_count, 2) + 
    poly(정류장_count_100, 2) + poly(grid_유소년인구_100, 
    2) + poly(grid_학원수_100, 2), data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-0.7901 -0.2749 -0.1887  0.3435  1.6174 

Coefficients:
                              Estimate Std. Error t value Pr(>|t|)    
(Intercept)                    0.29486    0.02499  11.801  < 2e-16 ***
poly(어린이집_count, 2)1       1.71864    0.59379   2.894  0.00405 ** 
poly(어린이집_count, 2)2       0.53121    0.48673   1.091  0.27588    
poly(정류장_count_100, 2)1     1.35539    0.47016   2.883  0.00420 ** 
poly(정류장_count_100, 2)2     0.54269    0.47421   1.144  0.25327    
poly(grid_유소년인구_100, 2)1 -0.89849    0.55705  -1.613  0.10769    
poly(grid_유소년인구_100, 2)2  0.05992    0.50979   0.118  0.90651    
poly(grid_학원수_100, 2)1      1.59423    0.49335   3.231  0.00135 ** 
poly(grid_학원수_100, 2)2      0.25439    0.47415   0.537  0.59195    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05

- 결정계수 0.07 >> 0.10. 으로 증가, 유의하지 않은 항 제거하기

In [22]:
# (2) 종속변수 logarithm applying (x) / 설명변수 2차항까지 fitting
lm.final3 <- lm(accident_cnt ~ poly(어린이집_count,2) + poly(정류장_count_100,2)  + poly(grid_유소년인구_100,2) + poly(grid_학원수_100,2),acc1)
summary(lm.final3)


Call:
lm(formula = accident_cnt ~ poly(어린이집_count, 2) + poly(정류장_count_100, 
    2) + poly(grid_유소년인구_100, 2) + poly(grid_학원수_100, 
    2), data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.2466 -0.4702 -0.3547  0.3536  6.9212 

Coefficients:
                              Estimate Std. Error t value Pr(>|t|)    
(Intercept)                    1.55652    0.05736  27.134  < 2e-16 ***
poly(어린이집_count, 2)1       3.91338    1.36318   2.871  0.00435 ** 
poly(어린이집_count, 2)2       0.81424    1.11740   0.729  0.46670    
poly(정류장_count_100, 2)1     3.31157    1.07937   3.068  0.00233 ** 
poly(정류장_count_100, 2)2     2.91061    1.08866   2.674  0.00787 ** 
poly(grid_유소년인구_100, 2)1 -2.13117    1.27884  -1.666  0.09655 .  
poly(grid_유소년인구_100, 2)2  0.65963    1.17035   0.564  0.57339    
poly(grid_학원수_100, 2)1      2.69714    1.13260   2.381  0.01780 *  
poly(grid_학원수_100, 2)2      0.58796    1.08852   0.540  0.58945    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 

- 결정계수 0.07 >> 0.09 증가 
- 유의해 보이는 2차항은 정류장_count_100

In [56]:
# 유의한 2차항인 정류장_count_100 만 2차항 추가
lm.final4 <- lm(accident_cnt~어린이집_count+poly(정류장_count_100,2)+grid_유소년인구_100+grid_학원수_100,acc1)
summary(lm.final4)


Call:
lm(formula = accident_cnt ~ 어린이집_count + poly(정류장_count_100, 
    2) + grid_유소년인구_100 + grid_학원수_100, data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.3196 -0.4902 -0.3316  0.3387  6.9243 

Coefficients:
                             Estimate Std. Error t value Pr(>|t|)    
(Intercept)                 1.3708730  0.0843843  16.246  < 2e-16 ***
어린이집_count              0.0386902  0.0131156   2.950  0.00340 ** 
poly(정류장_count_100, 2)1  3.2762124  1.0749088   3.048  0.00249 ** 
poly(정류장_count_100, 2)2  2.8638193  1.0851370   2.639  0.00870 ** 
grid_유소년인구_100        -0.0014535  0.0008102  -1.794  0.07370 .  
grid_학원수_100             0.0118409  0.0049760   2.380  0.01788 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.063 on 339 degrees of freedom
Multiple R-squared:  0.09479,	Adjusted R-squared:  0.08144 
F-statistic:   7.1 on 5 and 339 DF,  p-value: 2.483e-06


- 설명변수는 줄어들었지만 비슷한 수준의 결정계수를 유지
- 마찬가지로 grid_유소년인구_100은 약간 벗어나지만 계속 가져가기로

In [25]:
# 3차항도 유의한지, 넣어서 model fitting
lm.final5 <- lm(accident_cnt~어린이집_count+poly(정류장_count_100,3)+grid_유소년인구_100+grid_학원수_100,acc1)
summary(lm.final5)


Call:
lm(formula = accident_cnt ~ 어린이집_count + poly(정류장_count_100, 
    3) + grid_유소년인구_100 + grid_학원수_100, data = acc1)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.0121 -0.4836 -0.3363  0.3419  6.2354 

Coefficients:
                             Estimate Std. Error t value Pr(>|t|)    
(Intercept)                 1.3885717  0.0839820  16.534  < 2e-16 ***
어린이집_count              0.0353486  0.0130740   2.704  0.00720 ** 
poly(정류장_count_100, 3)1  3.2770085  1.0661680   3.074  0.00229 ** 
poly(정류장_count_100, 3)2  2.8296220  1.0763955   2.629  0.00896 ** 
poly(정류장_count_100, 3)3  2.7232356  1.0615232   2.565  0.01074 *  
grid_유소년인구_100        -0.0014611  0.0008036  -1.818  0.06991 .  
grid_학원수_100             0.0120265  0.0049361   2.436  0.01535 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.054 on 338 degrees of freedom
Multiple R-squared:  0.1121,	Adjusted R-squared:  0.09632 
F-statistic: 7.111 on 6 and 338 DF,  p-value: 

- 3차항도 유의
- grid_유소년인구_100 설명변수의 유의성 또한 증가
- 결정계수 0.09 >> 0.11 증가
__>> 예비 model adopt__

In [ ]:
# Cohen, J, Statistical Power Analysis for the Behavioral Sciences(2nd Ed.)
# Coefficient of R - Squared
# 작은 크기 : 2%
# 중간 크기 : 13%
# 큰   크기 : 35%

# - Model Validation
- random seed는 2021로 부여
### LOOCV Approach

In [26]:
library(boot)

In [29]:
set.seed(2021)
glm.fit = glm(accident_cnt~어린이집_count+poly(정류장_count_100,3)+grid_유소년인구_100+grid_학원수_100,data=acc1)
cv.err = cv.glm(acc1,glm.fit)

In [30]:
cv.err$delta[1] # mse 1.2219

[1] 1.221952

In [31]:
summary(cv.err)

      Length Class  Mode   
call    3    -none- call   
K       1    -none- numeric
delta   2    -none- numeric
seed  626    -none- numeric

- LOOCV MSE : 1.2219

-----------------------------------------------------------------

### K-Fold validation

In [32]:
set.seed(2021)
cv.glm(acc1,glm.fit,K=5)$delta[1]

[1] 1.180624

In [33]:
set.seed(2021)
cv.glm(acc1,glm.fit,K=10)$delta[1]

[1] 1.225564

- 5-fold validation >> MSE : 1.18062
- 10-fold validation >> MSE : 1.22556

-----------------------------------------------------------------

### Validation Set Approach

In [43]:
set.seed(2021)
train = sort(sample(1:dim(acc1)[1],dim(acc1)[1]*0.8))       # 8대2 비율로 train set의 index를 random sampling
test = setdiff(1:dim(acc1)[1],train)

y_test = acc1$accident_cnt[test]      # test set
x_test = acc1[test,-1]

lm.fit = lm(accident_cnt~어린이집_count+poly(정류장_count_100,3)+grid_유소년인구_100+grid_학원수_100,data=acc1,subset=train)
yhat = predict(lm.fit,x_test)         # x_test를 새로운 data로 predict
mean((y_test-yhat)^2)

[1] 1.150482

In [44]:
summary(lm.fit)


Call:
lm(formula = accident_cnt ~ 어린이집_count + poly(정류장_count_100, 
    3) + grid_유소년인구_100 + grid_학원수_100, data = acc1, 
    subset = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.6415 -0.4946 -0.3323  0.3211  5.6264 

Coefficients:
                             Estimate Std. Error t value Pr(>|t|)    
(Intercept)                 1.3952412  0.0944457  14.773  < 2e-16 ***
어린이집_count              0.0269371  0.0144406   1.865  0.06322 .  
poly(정류장_count_100, 3)1  3.9062420  1.2167671   3.210  0.00149 ** 
poly(정류장_count_100, 3)2  3.6515773  1.2322117   2.963  0.00331 ** 
poly(정류장_count_100, 3)3  3.6223452  1.1851050   3.057  0.00246 ** 
grid_유소년인구_100        -0.0010818  0.0008489  -1.274  0.20367    
grid_학원수_100             0.0133564  0.0058733   2.274  0.02375 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.054 on 269 degrees of freedom
Multiple R-squared:  0.1333,	Adjusted R-squared:  0.1139 
F-statistic: 6.892 on 6 an

- train/test(8:2) MSE : 1.1504

### Model의 성능은 
- Validation set / K-Fold / LOOCV 순
- Validation set 모델의 R-squared 도 0.133 으로 상승

In [57]:
# 유의하지 않은 grid_유소년인구_100 변수 제거
set.seed(2021)
train = sort(sample(1:dim(acc1)[1],dim(acc1)[1]*0.8))       # 5대5 비율로 train set의 index를 random sampling
test = setdiff(1:dim(acc1)[1],train)

y_test = acc1$accident_cnt[test]      # test set
x_test = acc1[test,-1]                # acc1의 첫번째 열이 accident_cnt

lm.fit = lm(accident_cnt~어린이집_count+poly(정류장_count_100,3)+grid_학원수_100,data=acc1,subset=train)
yhat = predict(lm.fit,x_test)         # x_test를 새로운 data로 predict
mean((y_test-yhat)^2)

[1] 1.176553

In [58]:
summary(lm.fit)


Call:
lm(formula = accident_cnt ~ 어린이집_count + poly(정류장_count_100, 
    3) + grid_학원수_100, data = acc1, subset = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.6717 -0.4980 -0.3394  0.3652  5.5908 

Coefficients:
                           Estimate Std. Error t value Pr(>|t|)    
(Intercept)                1.378417   0.093626  14.723  < 2e-16 ***
어린이집_count             0.017999   0.012637   1.424  0.15552    
poly(정류장_count_100, 3)1 3.741189   1.211250   3.089  0.00222 ** 
poly(정류장_count_100, 3)2 3.905168   1.217438   3.208  0.00150 ** 
poly(정류장_count_100, 3)3 3.647427   1.186309   3.075  0.00232 ** 
grid_학원수_100            0.013818   0.005869   2.354  0.01927 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.055 on 270 degrees of freedom
Multiple R-squared:  0.128,	Adjusted R-squared:  0.1119 
F-statistic: 7.928 on 5 and 270 DF,  p-value: 5.488e-07


- MSE값 증가와 함께, 어린이집_count의 유의성도 떨어짐
    - 100m의 유소년 인구를 반경 300m로 넓혀봄 (grid_유소년인구 변수 추가)

In [59]:
# 유의하지 않은 grid_유소년인구_100 변수 제거
set.seed(2021)
train = sort(sample(1:dim(acc1)[1],dim(acc1)[1]*0.8))       # 5대5 비율로 train set의 index를 random sampling
test = setdiff(1:dim(acc1)[1],train)

y_test = acc1$accident_cnt[test]      # test set
x_test = acc1[test,-1]                # acc1의 첫번째 열이 accident_cnt

lm.fit = lm(accident_cnt~어린이집_count+poly(정류장_count_100,3)+grid_유소년인구+grid_학원수_100,data=acc1,subset=train)
yhat = predict(lm.fit,x_test)         # x_test를 새로운 data로 predict
mean((y_test-yhat)^2)

[1] 1.169091

In [60]:
summary(lm.fit)


Call:
lm(formula = accident_cnt ~ 어린이집_count + poly(정류장_count_100, 
    3) + grid_유소년인구 + grid_학원수_100, data = acc1, 
    subset = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.6004 -0.4989 -0.3343  0.2805  5.6036 

Coefficients:
                             Estimate Std. Error t value Pr(>|t|)    
(Intercept)                 1.4041522  0.0945811  14.846  < 2e-16 ***
어린이집_count              0.0385661  0.0176151   2.189  0.02943 *  
poly(정류장_count_100, 3)1  3.9446651  1.2133871   3.251  0.00130 ** 
poly(정류장_count_100, 3)2  3.5823219  1.2287240   2.915  0.00385 ** 
poly(정류장_count_100, 3)3  3.5698312  1.1833100   3.017  0.00280 ** 
grid_유소년인구            -0.0006546  0.0003919  -1.670  0.09605 .  
grid_학원수_100             0.0132500  0.0058594   2.261  0.02454 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.052 on 269 degrees of freedom
Multiple R-squared:  0.137,	Adjusted R-squared:  0.1177 
F-statistic: 7.115 on 6 and 269

- grid 유소년 인구의 유의성은 약간 떨어지지만 다른 변수의 유의성 증가
- model의 결정계수 또한 0.137로 증가 및 MSE 1.17 --> 1.16으로 소폭 감소
### 최종 모델로 채택

In [ ]:
## 최종 모델
yhat = 1.4041522 + 0.0385661* 어린이집_count + 3.9446651*(정류장_count_100)+ 
        3.5823219*(정류장_count_100)^2 3.5698312*(정류장_count_100)^3 
        -0.0006546*grid_유소년인구 +0.0132500*grid_학원수_100

In [63]:
yhat <- predict(lm.fit,acc1[-1])

In [64]:
print(yhat)

        1         2         3         4         5         6         7         8 
1.3901837 1.2562328 1.2562328 1.2562328 1.5310252 1.3137265 1.2957140 1.2523051 
        9        10        11        12        13        14        15        16 
1.2947990 1.2490321 1.3196181 1.3901837 1.2947990 1.4003197 2.1814919 1.3572480 
       17        18        19        20        21        22        23        24 
1.5692109 1.1938753 1.5364662 1.2186027 1.2980494 1.3333651 1.2947990 1.0547252 
       25        26        27        28        29        30        31        32 
1.3038510 1.7238557 1.2562328 1.4974998 1.2878944 1.3147337 1.4338741 1.4502958 
       33        34        35        36        37        38        39        40 
1.2562328 1.5310252 1.5310252 1.5459615 2.1766498 1.1939316 1.2842688 1.3771118 
       41        42        43        44        45        46        47        48 
1.1741241 1.3719312 1.1456172 1.3745643 2.4242174 1.4892069 1.2565741 1.6215445 
       49        50        5

In [65]:
acc1['score'] = yhat

In [66]:
head(acc1)

accident_cnt 어린이집_count 유치원_count 초등학교_count 방지턱_count
1 1            0              0            0              0           
2 1            0              0            0              0           
3 1            0              0            0              0           
4 1            0              0            0              0           
5 1            0              0            0              0           
6 1            2              1            0              0           
  불법주정차_cam_count 정류장_count_100 정류장_count_300 grid_유동인구
1 0                    2                2                43.435676    
2 0                    0                2                 9.633417    
3 0                    0                2                 0.215500    
4 0                    0                0                 0.662000    
5 0                    1                6                 8.104333    
6 0                    0                0                 2.751254    
  grid_유소년인구 ⋯ 유치원_count_100 초등학교_count_100 방지턱_count_100
1  0              ⋯ 0                0                  0               
2  0              ⋯ 0                0                  0               
3  0              ⋯ 0                0                  0               
4  0              ⋯ 0                0                  0               
5  0              ⋯ 0                0                  0               
6 30              ⋯ 0                0                  0               
  grid_유소년인구_100 grid_학원수_100 어린이집_count_200 유치원_count_200
1  0                  0               0                  0               
2  0                  0               0                  0               
3  0                  0               0                  0               
4  0                  0               0                  0               
5  0                  0               0                  0               
6 30                  0               0                  1               
  초등학교_count_200 방지턱_count_200 score   
1 0                  0                1.390184
2 0                  0                1.256233
3 0                  0                1.256233
4 0                  0                1.256233
5 0                  0                1.531025
6 0                  0                1.313727

In [68]:
write.csv(acc1,'acc1_model_score.csv')